File: ts/datastructure/otterop/datastructure/ListIterator.ts
import { OOPIterator } from '@otterop/lang/OOPIterator';
import { List } from './List';

/** @internal */
export class ListIterator<T> implements OOPIterator<T> {

    #list : List<T>;

    #index : number;

    /** @internal */
    constructor(list : List<T>) {
        this.#list = list;
        this.#index = 0;
    }

    public hasNext() : boolean {
        return this.#index < this.#list.size();
    }

    public next() : T {
        let ret : T = this.#list.get(this.#index);
        this.#index++;
        return ret;
    }
}

File: python/datastructure/otterop/datastructure/_list_iterator.py

class ListIterator:
        
    def __init__(self, list):
        self._list = list
        self._index = 0

    def has_next(self):
        return self._index < self._list.size()

    def next(self):
        ret = self._list.get(self._index)
        self._index += 1
        return ret
